File: platform/polyglot/confluence_api/mjs/src/utils/cql-builder.mjs
/**
 * @module utils/cql-builder
 * @description Fluent CQL (Confluence Query Language) query builder.
 *
 * Confluence Data Center uses CQL for searching content, similar to Jira's JQL.
 * This builder provides a safe, composable API for constructing CQL queries
 * without manual string concatenation or escaping concerns.
 *
 * CQL reference: https://developer.atlassian.com/server/confluence/advanced-searching-using-cql/
 *
 * Supported operators:
 * - `=`, `!=` — exact match / not equal
 * - `~`, `!~` — contains / does not contain (full-text)
 * - `in`, `not in` — list membership
 * - `is null`, `is not null` — null checks
 *
 * Logical operators: `AND`, `OR`, `NOT`
 * Ordering: `ORDER BY field ASC|DESC`
 *
 * @example
 * import { cql, CQLBuilder } from './cql-builder.mjs';
 *
 * // Simple query
 * const query = cql('space').equals('DEV').and().field('type').equals('page').build();
 * // => 'space = "DEV" AND type = "page"'
 *
 * // Complex query with ordering
 * const query = new CQLBuilder()
 *   .field('space').equals('DEV')
 *   .and()
 *   .field('type').equals('page')
 *   .and()
 *   .field('label').equals('architecture')
 *   .orderBy('lastModified', 'DESC')
 *   .build();
 * // => 'space = "DEV" AND type = "page" AND label = "architecture" ORDER BY lastModified DESC'
 */

/**
 * Escape a CQL string value by wrapping in double quotes and escaping
 * embedded double quotes and backslashes.
 *
 * @param {string} value - The raw string value to escape.
 * @returns {string} The escaped and quoted string.
 * @private
 */
function _escapeValue(value) {
  const escaped = String(value).replace(/\\/g, '\\\\').replace(/"/g, '\\"');
  return `"${escaped}"`;
}

/**
 * Fluent CQL query builder for Confluence content search.
 *
 * Builds CQL query strings incrementally through a chainable API.
 * Call `.build()` to produce the final query string.
 *
 * @example
 * const builder = new CQLBuilder();
 * const query = builder
 *   .field('space').equals('DEV')
 *   .and()
 *   .field('ancestor').equals('12345')
 *   .orderBy('created', 'ASC')
 *   .build();
 */
export class CQLBuilder {
  constructor() {
    /** @private @type {string[]} */
    this._parts = [];
    /** @private @type {string|null} */
    this._currentField = null;
    /** @private @type {string|null} */
    this._orderByClause = null;
  }

  /**
   * Set the current field for the next condition.
   *
   * @param {string} fieldName - CQL field name (e.g. 'space', 'type', 'label', 'ancestor').
   * @returns {CQLBuilder} This builder instance for chaining.
   *
   * @example
   * builder.field('space').equals('DEV');
   */
  field(fieldName) {
    this._currentField = fieldName;
    return this;
  }

  /**
   * Add an equality condition: `field = "value"`.
   *
   * @param {string} value - The value to match.
   * @returns {CQLBuilder} This builder instance for chaining.
   */
  equals(value) {
    this._assertField('equals');
    this._parts.push(`${this._currentField} = ${_escapeValue(value)}`);
    this._currentField = null;
    return this;
  }

  /**
   * Add a not-equal condition: `field != "value"`.
   *
   * @param {string} value - The value to exclude.
   * @returns {CQLBuilder} This builder instance for chaining.
   */
  notEquals(value) {
    this._assertField('notEquals');
    this._parts.push(`${this._currentField} != ${_escapeValue(value)}`);
    this._currentField = null;
    return this;
  }

  /**
   * Add a contains condition: `field ~ "value"` (full-text search).
   *
   * @param {string} value - The text to search for.
   * @returns {CQLBuilder} This builder instance for chaining.
   */
  contains(value) {
    this._assertField('contains');
    this._parts.push(`${this._currentField} ~ ${_escapeValue(value)}`);
    this._currentField = null;
    return this;
  }

  /**
   * Add a not-contains condition: `field !~ "value"`.
   *
   * @param {string} value - The text to exclude.
   * @returns {CQLBuilder} This builder instance for chaining.
   */
  notContains(value) {
    this._assertField('notContains');
    this._parts.push(`${this._currentField} !~ ${_escapeValue(value)}`);
    this._currentField = null;
    return this;
  }

  /**
   * Add an IN condition: `field in ("a", "b", "c")`.
   *
   * @param {string[]} values - The list of values to match.
   * @returns {CQLBuilder} This builder instance for chaining.
   */
  inList(values) {
    this._assertField('inList');
    const escaped = values.map(_escapeValue).join(', ');
    this._parts.push(`${this._currentField} in (${escaped})`);
    this._currentField = null;
    return this;
  }

  /**
   * Add a NOT IN condition: `field not in ("a", "b", "c")`.
   *
   * @param {string[]} values - The list of values to exclude.
   * @returns {CQLBuilder} This builder instance for chaining.
   */
  notInList(values) {
    this._assertField('notInList');
    const escaped = values.map(_escapeValue).join(', ');
    this._parts.push(`${this._currentField} not in (${escaped})`);
    this._currentField = null;
    return this;
  }

  /**
   * Add an IS NOT NULL condition: `field is not null`.
   *
   * @returns {CQLBuilder} This builder instance for chaining.
   */
  isNotNull() {
    this._assertField('isNotNull');
    this._parts.push(`${this._currentField} is not null`);
    this._currentField = null;
    return this;
  }

  /**
   * Add an IS NULL condition: `field is null`.
   *
   * @returns {CQLBuilder} This builder instance for chaining.
   */
  isNull() {
    this._assertField('isNull');
    this._parts.push(`${this._currentField} is null`);
    this._currentField = null;
    return this;
  }

  /**
   * Add an AND logical operator between conditions.
   *
   * @returns {CQLBuilder} This builder instance for chaining.
   */
  and() {
    this._parts.push('AND');
    return this;
  }

  /**
   * Add an OR logical operator between conditions.
   *
   * @returns {CQLBuilder} This builder instance for chaining.
   */
  or() {
    this._parts.push('OR');
    return this;
  }

  /**
   * Add a NOT logical operator (prefix for the next condition).
   *
   * @returns {CQLBuilder} This builder instance for chaining.
   */
  not() {
    this._parts.push('NOT');
    return this;
  }

  /**
   * Set the ORDER BY clause. Replaces any previously set ordering.
   *
   * @param {string} fieldName - Field to order by (e.g. 'created', 'lastModified', 'title').
   * @param {'ASC'|'DESC'} [direction='ASC'] - Sort direction.
   * @returns {CQLBuilder} This builder instance for chaining.
   *
   * @example
   * builder.field('type').equals('page').orderBy('lastModified', 'DESC').build();
   * // => 'type = "page" ORDER BY lastModified DESC'
   */
  orderBy(fieldName, direction = 'ASC') {
    const dir = direction.toUpperCase() === 'DESC' ? 'DESC' : 'ASC';
    this._orderByClause = `ORDER BY ${fieldName} ${dir}`;
    return this;
  }

  /**
   * Build the final CQL query string.
   *
   * Joins all accumulated parts with spaces and appends the ORDER BY clause
   * if one was specified.
   *
   * @returns {string} The complete CQL query string.
   * @throws {Error} If no conditions have been added.
   *
   * @example
   * const query = new CQLBuilder()
   *   .field('space').equals('DEV')
   *   .and()
   *   .field('type').equals('page')
   *   .build();
   * // => 'space = "DEV" AND type = "page"'
   */
  build() {
    if (this._parts.length === 0) {
      throw new Error('CQL query is empty — add at least one condition before building');
    }

    let query = this._parts.join(' ');
    if (this._orderByClause) {
      query += ` ${this._orderByClause}`;
    }
    return query;
  }

  /**
   * Assert that a current field has been set before applying an operator.
   *
   * @param {string} operatorName - Name of the operator being applied (for error messages).
   * @throws {Error} If no field has been set via `.field()`.
   * @private
   */
  _assertField(operatorName) {
    if (!this._currentField) {
      throw new Error(
        `CQLBuilder: call .field(name) before .${operatorName}()`,
      );
    }
  }
}

/**
 * Shortcut factory: create a CQLBuilder and set the first field in one call.
 *
 * @param {string} fieldName - The initial field name.
 * @returns {CQLBuilder} A new CQLBuilder with the field already set.
 *
 * @example
 * const query = cql('space').equals('DEV')
 *   .and()
 *   .field('type').equals('page')
 *   .build();
 * // => 'space = "DEV" AND type = "page"'
 */
export function cql(fieldName) {
  return new CQLBuilder().field(fieldName);
}

File: platform/polyglot/confluence_api/py/src/confluence_api/utils/cql_builder.py
"""
CQL (Confluence Query Language) query builder with fluent API.

Usage:
    from confluence_api.utils.cql_builder import cql, CQLBuilder

    query = (
        cql("type").equals("page")
        .and_()
        .field("space").equals("DEV")
        .and_()
        .field("title").contains("architecture")
        .order_by("lastModified", "desc")
        .build()
    )
    # => 'type = "page" AND space = "DEV" AND title ~ "architecture" ORDER BY lastModified desc'
"""

from __future__ import annotations

from typing import Any


class CQLBuilder:
    """
    Fluent CQL query builder for Confluence search.

    CQL (Confluence Query Language) is used with the /content/search and
    /search endpoints. This builder provides a safe, composable way to
    construct CQL queries without manual string concatenation.
    """

    def __init__(self, field_name: str | None = None) -> None:
        """
        Initialize the CQL builder.

        Args:
            field_name: Optional initial field name to query against.
        """
        self._parts: list[str] = []
        self._current_field: str | None = field_name
        self._order_by: str | None = None

    def field(self, name: str) -> CQLBuilder:
        """
        Set the current field for the next comparison operation.

        Args:
            name: CQL field name (e.g. 'type', 'space', 'title', 'label',
                  'creator', 'lastModified', 'ancestor', 'content').

        Returns:
            Self for method chaining.
        """
        self._current_field = name
        return self

    def equals(self, value: str) -> CQLBuilder:
        """
        Add an equality condition: field = "value".

        Args:
            value: The value to compare against.

        Returns:
            Self for method chaining.
        """
        self._assert_field()
        self._parts.append(f'{self._current_field} = {_quote(value)}')
        self._current_field = None
        return self

    def not_equals(self, value: str) -> CQLBuilder:
        """
        Add a not-equal condition: field != "value".

        Args:
            value: The value to compare against.

        Returns:
            Self for method chaining.
        """
        self._assert_field()
        self._parts.append(f'{self._current_field} != {_quote(value)}')
        self._current_field = None
        return self

    def contains(self, value: str) -> CQLBuilder:
        """
        Add a contains/matches condition: field ~ "value".

        Uses CQL's ~ operator for text search matching.

        Args:
            value: The text to search for.

        Returns:
            Self for method chaining.
        """
        self._assert_field()
        self._parts.append(f'{self._current_field} ~ {_quote(value)}')
        self._current_field = None
        return self

    def not_contains(self, value: str) -> CQLBuilder:
        """
        Add a not-contains condition: field !~ "value".

        Args:
            value: The text to exclude.

        Returns:
            Self for method chaining.
        """
        self._assert_field()
        self._parts.append(f'{self._current_field} !~ {_quote(value)}')
        self._current_field = None
        return self

    def in_list(self, values: list[str]) -> CQLBuilder:
        """
        Add an IN condition: field IN ("val1", "val2", ...).

        Args:
            values: List of values for the IN clause.

        Returns:
            Self for method chaining.
        """
        self._assert_field()
        quoted = ", ".join(_quote(v) for v in values)
        self._parts.append(f'{self._current_field} IN ({quoted})')
        self._current_field = None
        return self

    def not_in_list(self, values: list[str]) -> CQLBuilder:
        """
        Add a NOT IN condition: field NOT IN ("val1", "val2", ...).

        Args:
            values: List of values for the NOT IN clause.

        Returns:
            Self for method chaining.
        """
        self._assert_field()
        quoted = ", ".join(_quote(v) for v in values)
        self._parts.append(f'{self._current_field} NOT IN ({quoted})')
        self._current_field = None
        return self

    def is_not_null(self) -> CQLBuilder:
        """
        Add an IS NOT NULL condition: field IS NOT NULL.

        Returns:
            Self for method chaining.
        """
        self._assert_field()
        self._parts.append(f'{self._current_field} IS NOT NULL')
        self._current_field = None
        return self

    def is_null(self) -> CQLBuilder:
        """
        Add an IS NULL condition: field IS NULL.

        Returns:
            Self for method chaining.
        """
        self._assert_field()
        self._parts.append(f'{self._current_field} IS NULL')
        self._current_field = None
        return self

    def and_(self) -> CQLBuilder:
        """
        Add AND logical operator between conditions.

        Returns:
            Self for method chaining.
        """
        self._parts.append("AND")
        return self

    def or_(self) -> CQLBuilder:
        """
        Add OR logical operator between conditions.

        Returns:
            Self for method chaining.
        """
        self._parts.append("OR")
        return self

    def not_(self) -> CQLBuilder:
        """
        Add NOT logical operator before the next condition.

        Returns:
            Self for method chaining.
        """
        self._parts.append("NOT")
        return self

    def order_by(self, field: str, direction: str = "asc") -> CQLBuilder:
        """
        Set the ORDER BY clause.

        Args:
            field: Field name to sort by (e.g. 'lastModified', 'created', 'title').
            direction: Sort direction, either 'asc' or 'desc' (default 'asc').

        Returns:
            Self for method chaining.
        """
        direction = direction.lower()
        if direction not in ("asc", "desc"):
            raise ValueError(f"Invalid sort direction: {direction!r}. Must be 'asc' or 'desc'.")
        self._order_by = f"{field} {direction}"
        return self

    def build(self) -> str:
        """
        Build and return the final CQL query string.

        Returns:
            The complete CQL query string.

        Raises:
            ValueError: If no conditions have been added.
        """
        if not self._parts:
            raise ValueError("Cannot build empty CQL query. Add at least one condition.")

        query = " ".join(self._parts)

        if self._order_by:
            query += f" ORDER BY {self._order_by}"

        return query

    def _assert_field(self) -> None:
        """Assert that a current field is set for the next comparison."""
        if not self._current_field:
            raise ValueError(
                "No field set. Call .field('name') before a comparison method, "
                "or use cql('field_name') to initialize with a field."
            )

    def __repr__(self) -> str:
        try:
            return f"CQLBuilder(query={self.build()!r})"
        except ValueError:
            return "CQLBuilder(query=<empty>)"


def cql(field_name: str) -> CQLBuilder:
    """
    Shortcut constructor for CQLBuilder with an initial field.

    Args:
        field_name: The first field to query against.

    Returns:
        A new CQLBuilder instance with the field pre-set.

    Example:
        >>> cql("type").equals("page").and_().field("space").equals("DEV").build()
        'type = "page" AND space = "DEV"'
    """
    return CQLBuilder(field_name)


def _quote(value: str) -> str:
    """
    Quote a CQL string value with double quotes, escaping inner double quotes.

    Args:
        value: The string value to quote.

    Returns:
        The quoted string suitable for CQL.
    """
    escaped = value.replace("\\", "\\\\").replace('"', '\\"')
    return f'"{escaped}"'
